File: exercises/RB101_small_problems/easy_1/how_many.rb
# Solution using PEDAC

# We understand the problem:
# Write a method that counts the number of occurrences of each element in a given array.

# Test Case
  # car => 4
  # truck => 3
  # SUV => 1
  # motorcycle => 2

# Data Structure:
  # input: array of cars
  # output: how many different types of cars

# Algorithm:
  # create method `count_occurences(array)`
    # create hash with default value as 0
    # iterate over array and increment each car type
    # return hash

# Code with intend:

def count_occurrences(array)
  vehicles = Hash.new(0)

  car_count = array.select do |value|
    # further exploration - case sensitive values
    value.upcase! if value == 'suv'

    vehicles[value] += 1
  end
  
  vehicles.each {|key, value| puts "#{key} => #{value}"}
    
end

vehicles = [
  'car', 'car', 'truck', 'car', 'suv', 'truck',
  'motorcycle', 'motorcycle', 'car', 'truck'
]

count_occurrences(vehicles)

File: exercises/JS210_Small_Problems/Easy_4/howMany.js
function countOccurrences(list) {
  const counter = list.reduce((count, item) => {
    count[item] ? count[item] += 1 : count[item] = 1;
    return count;
  }, {});

  logOccurrences(counter);
  // console.table(counter)
}

function logOccurrences(occurrences) {
  for (let item in occurrences) {
    console.log(`${item} => ${occurrences[item]}`);
  }
}

const vehicles = ['car', 'car', 'truck', 'car', 'SUV', 'truck',
                'motorcycle', 'suv', 'motorcycle', 'car', 'truck'];

countOccurrences(vehicles);

// console output
// car => 4
// truck => 3
// SUV => 1
// motorcycle => 2
// suv => 1